src/iframe/storage/unpartitioned-storage/unpartitioned-storage.ts

```diff
@@ -525,25 +197,26 @@ export class EnhancedStorage implements Storage {
    */
   protected dispatchUnpartitionedStateStatusChange(
     unpartitionedStateStatusOrError: UnpartitionedStateStatus | Error,
-  ): void {
+  ): UnpartitionedStateStatus {
     const unpartitionedStateStatus = isError(unpartitionedStateStatusOrError)
       ? "error"
       : unpartitionedStateStatusOrError;
-    const error = isError(unpartitionedStateStatusOrError) ? unpartitionedStateStatusOrError : undefined;
 
-    log(LOG_GROUP.STORAGE, `Unpartitioned state access for = ${unpartitionedStateStatus} (${this.storageType})`);
+    if (unpartitionedStateStatus === "error") {
+      this.error = isError(unpartitionedStateStatusOrError)
+        ? unpartitionedStateStatusOrError
+        : new Error("Unexpected unpartitioned state error");
+    } else {
+      this.error = null;
+    }
+
+    log(LOG_GROUP.STORAGE, `Unpartitioned state access for ${this.storageType} = ${unpartitionedStateStatus}`);
+
+    setUnpartitionedStateStatus(unpartitionedStateStatus, this.error);
```

Comment:
The `UNPARTITIONED_STATE_STATUS_CHANGE_EVENT` event has been replaced with callbacks. `setUnpartitionedStateStatus()` will invoke them. With this change, as soon as you add a callback, it is invoked with the most recent status, fixing a race condition we had before where if we started listening after the status had changed, we won't be notified, so Brave could sometimes show the UI as if it had access to unpartitioned state, when it actually didn't.